File: nodejs/coinbasepro-listener.js
const Gdax = require('gdax');
var _ = require('underscore');
const BigNumber = require('bignumber.js');

// Parameters
var maxEntryCount = 3
var maxVolumeCount = Infinity
var pairs = ['BCH/BTC', 'BTC/EUR', 'LTC/EUR', 'BTC/USD', 'BTC/EUR', 'ETH/USD',
'ETH/EUR', 'BCH/EUR', 'ETH/BTC', 'BCH/USD']

const publicClient = new Gdax.PublicClient();
const orderbookSync = new Gdax.OrderbookSync(pairs.map(x => x.replace('/','-')));
var asksConsolidated_old = new Array()
var bidsConsolidated_old = new Array()

var kafka = require('kafka-node'),
    HighLevelProducer = kafka.HighLevelProducer,
    client = new kafka.KafkaClient({kafkaHost: "kafka.cryptoindex.me:9092"}),
    producer = new HighLevelProducer(client);

// producer.on('ready', function () {
//     console.log("Kafka producer is ready");
//     // producerReady = true;
// });
      
// producer.on('error', function (err) {
//   console.error("Problem with producing to Kafka " + err);
// })

function getConsolidatedOrderbook(entries) {
    var orderbook = new Array()
    var price = entries[0].price
    var size = new BigNumber(0)    
    var sizeAccumulator = 0
    for (var i = 0; i < entries.length; i++) {

        if (entries[i].price.isEqualTo(price)) {
            size = size.plus(entries[i].size)
        }
        else {
            orderbook.push([price, size])
            sizeAccumulator += size.toNumber()
            if (orderbook.length >= maxEntryCount || sizeAccumulator >= maxVolumeCount) {
                return orderbook
            }
            price = entries[i].price
            size = entries[i].size
        }
    }
}

var messageHandler = function (data) {
    try {

        if (_.isUndefined(data.product_id)) {
            return
        }
        orderbookState = orderbookSync.books[data.product_id].state()
        if (orderbookState.asks.length == 0 || orderbookState.bids.length == 0) {
            return
        }

        asksConsolidated = getConsolidatedOrderbook(orderbookState.asks)
        bidsConsolidated = getConsolidatedOrderbook(orderbookState.bids)

        if (!_.isEqual(asksConsolidated, asksConsolidated_old) || !_.isEqual(bidsConsolidated, bidsConsolidated_old)) {
            var delay = new Date().getTime() - (new Date(data.time).getTime());
            var symbol = data.product_id.replace("-", "/")

            var payload = {
                "exchange": "coinbasepro",
                "symbol": symbol,
                "data": {
                    "asks": asksConsolidated.map(x => [x[0].toNumber(),x[1].toNumber()]),
                    "bids": bidsConsolidated.map(x => [x[0].toNumber(),x[1].toNumber()])
                },
                "timestamp": new Date(data.time).getTime()
            };
                payloads = [
                    { topic: "orderbook", messages: JSON.stringify(JSON.stringify(payload)) },
                ];
                producer.send(payloads, function (err, data) {
                    if (err) throw err;
                    console.log(data);
                });

            console.log(data.product_id + " asks:" + asksConsolidated.toString() + ", bids:" + bidsConsolidated.toString() + " delay:" + delay.toString() + "ms");

            asksConsolidated_old = asksConsolidated
            bidsConsolidated_old = bidsConsolidated
        }
    }
    catch (err) {
        console.log('error: ' + err)
    }
}

orderbookSync.on('message', data => { messageHandler(data) })
orderbookSync.on('error', err => { console.log(err) });

File: src/coinbasepro-listener.py
import cbpro
from decimal import *
from logger import logger
import sys
import yaml
import json
from kafka import KafkaProducer
import dateutil.parser
import time
from multiprocessing import Process, Event
from CWMetrics import CWMetrics

# Parse config
with open(sys.argv[1], 'r') as config_file:
    config = yaml.load(config_file)


class CryptoArbOrderBook(cbpro.OrderBook):
    def __init__(self, maxEntryCount=10, timeLimiterSeconds=0.025, product_id='BTC-USD', log_to=None):
        self.maxEntryCount = maxEntryCount
        self.asksConsolidatedOld = []
        self.bidsConsolidatedOld = []
        self.timeLastRun = time.time()
        self.timeLimiterSeconds = timeLimiterSeconds
        self.metrics = CWMetrics(config['exchange']['name'])
        self.kafka_producer = KafkaProducer(bootstrap_servers=config['kafka']['address'],
                                            value_serializer=lambda v: json.dumps(v).encode('utf-8'))

        super().__init__(product_id=product_id, log_to=log_to)

    def on_message(self, message):
        try:
            logger.info(message)
            super().on_message(message)
        except Exception as err:
            logger.error("Error during calling cbpro.Orderbook on_message:" + str(err))
            return
                
        if (time.time() - self.timeLastRun) > self.timeLimiterSeconds:
            try:
                book = self.get_current_book()
                if len(book["asks"])>0 and len(book["bids"])>0:
                    asksConsolidated = self.getConsolidatedOrderbook(book["asks"], reverse=False)
                    bidsConsolidated = self.getConsolidatedOrderbook(book["bids"], reverse=True)
                    #print("asks:"+str(asksConsolidated)+", bids:"+str(bidsConsolidated))

                    if self.asksConsolidatedOld != asksConsolidated or self.bidsConsolidatedOld != bidsConsolidated:
                        payload = {}
                        payload['exchange'] = "coinbasepro"
                        payload['symbol'] = message['product_id'].replace('-','/')
                        payload['data'] = {}
                        payload['data']['asks'] = asksConsolidated
                        payload['data']['bids'] = bidsConsolidated
                        dt = dateutil.parser.parse(message['time'])
                        payload['timestamp'] = int(time.mktime(dt.timetuple()) * 1000 + dt.microsecond / 1000)

                        p = json.dumps(payload, separators=(',', ':'))
                        self.kafka_producer.send(config['kafka']['topic'], p)

                        logger.info("Received " + payload['symbol'] + " prices from coinbasepro")
                        self.metrics.put(payload['timestamp'])

                        self.asksConsolidatedOld = asksConsolidated
                        self.bidsConsolidatedOld = bidsConsolidated

                    if book["asks"][0][0] <= book["bids"][-1][0]:
                        logger.error("Bid higher than ask")
                        self.metrics.putError()
            except Exception as err:
                logger.error("Error during message processing:" + str(err))
            finally:
                self.timeLastRun = time.time()

    def getConsolidatedOrderbook(self, entries, reverse=False):
        orderbook = []
        
        size = Decimal(0)
        sizeAccumulator = 0

        if reverse:
            price = entries[-1][0]
            entries = reversed(entries)
        else:
            price = entries[0][0]

        for entry in entries:
            if entry[0] == price:
                size += entry[1]
            else:
                orderbook.append([float(price), float(size)])
                sizeAccumulator += size
                if len(orderbook) >= self.maxEntryCount:
                    break
                price = entry[0]
                size = entry[1]
        
        return orderbook
        
def CryptoArbOrderBookProcess(pair,stopProcessesEvent):
    orderbook=CryptoArbOrderBook(product_id=pair.replace('/', '-'))
    orderbook.start()
    stopProcessesEvent.wait()
    orderbook.close()

pairs = config['exchange']['symbols']


while True:
    stopProcessesEvent = Event()
    processes = [Process(target=CryptoArbOrderBookProcess, args=(pair, stopProcessesEvent)) for pair in pairs]

    # start order books
    for process in processes:
        process.daemon = True
        process.start()

    time.sleep(10)

    stopProcessesEvent.set()
    for process in processes:
        process.join()

logger.info("coinbaseproWebsocket exited normally. Bye.")


